File: app/src/main/java/remix/myplayer/util/ImageUtil.java
package remix.myplayer.util;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import remix.myplayer.application.APlayerApplication;

/**
 * Created by taeja on 16-3-23.
 */
public class ImageUtil {
    private static ImageUtil mInstance;
    public static ImageUtil getInstance(){
        return mInstance == null ? new ImageUtil() : mInstance;
    }

    /**
     * 获取图片大小
     * @param resid 图片资源id
     * @return 返回图片高度与宽度
     */
    public static int[] getImageSize(int resid){
        //获得图片参数
        BitmapFactory.Options options = new BitmapFactory.Options();
        //不为其分配内存
        options.inJustDecodeBounds = true;

        BitmapFactory.decodeResource(APlayerApplication.getContext().getResources(), resid,options);
        int imageHeight = options.outHeight;
        int imageWidth = options.outWidth;
        String imageType = options.outMimeType;
        return imageHeight > 0 && imageWidth > 0 ? new int[]{imageHeight,imageWidth} : null;
    }

    /**
     * @param originWidth 原始高度
     * @param originHeight 原始宽度
     * @param reqWidth 期望宽度
     * @param reqHeight 期望高度
     * @return 压缩比
     */
    public static int calculateInSampleSize(int originWidth,int originHeight,int reqWidth,int reqHeight){
        int inSampleSize = 1;
        if(originHeight > reqHeight || originWidth > reqWidth){
            final int heightRatio = Math.round((float)originHeight / reqHeight);
            final int widthRatio = Math.round((float)originWidth / reqWidth);
            inSampleSize = heightRatio < widthRatio ? heightRatio : widthRatio;
        }
        return inSampleSize;
    }

    /**
     * 将图片压缩到期望大小
     * @param reqHeight 期望高度
     * @param reqWidth 期望宽度
     * @param resid 资源id
     * @return 压缩后的bitmap
     */
    public static Bitmap decodeSampledBitmapFromResource(int reqHeight,int reqWidth,int resid){
        final BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeResource(APlayerApplication.getContext().getResources(),resid,options);

        options.inSampleSize = calculateInSampleSize(options.outWidth,options.outHeight,reqWidth,reqHeight);
        options.inJustDecodeBounds = false;

        return BitmapFactory.decodeResource(APlayerApplication.getContext().getResources(),resid,options);
    }
}

File: app/src/main/java/remix/myplayer/util/ImageUtil.kt
package remix.myplayer.util

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import remix.myplayer.data.model.audio.Song
import remix.myplayer.helper.AudioTagFile
import timber.log.Timber
import java.io.File

object ImageUtil {

  fun loadScaledBitmap(context: Context, uri: Uri, maxDimension: Int): Bitmap? {
    return try {
      // 1. Decode bounds only
      var input = context.contentResolver.openInputStream(uri) ?: return null
      val options = BitmapFactory.Options().apply {
        inJustDecodeBounds = true
      }
      BitmapFactory.decodeStream(input, null, options)
      input.close()

      // 2. Calculate inSampleSize
      var inSampleSize = 1
      if (options.outHeight > maxDimension || options.outWidth > maxDimension) {
        val halfHeight = options.outHeight / 2
        val halfWidth = options.outWidth / 2
        while ((halfHeight / inSampleSize) >= maxDimension && (halfWidth / inSampleSize) >= maxDimension) {
          inSampleSize *= 2
        }
      }

      // 3. Decode with inSampleSize
      input = context.contentResolver.openInputStream(uri) ?: return null
      options.inJustDecodeBounds = false
      options.inSampleSize = inSampleSize
      val bitmap = BitmapFactory.decodeStream(input, null, options)
      input.close()
      bitmap
    } catch (e: Exception) {
      Timber.e(e, "Failed to load scaled bitmap")
      null
    }
  }

  suspend fun extractAlbumArt(song: Song): ImageBitmap? = withContext(Dispatchers.IO) {
    if (!song.valid() || !song.isLocal()) {
      return@withContext null
    }

    try {
      val artwork = AudioTagFile.readFrontCover(File(song.data))
      if (artwork != null) {
        val artworkBinaryData = artwork.data
        val bitmap = BitmapFactory.decodeByteArray(
          artworkBinaryData,
          0,
          artworkBinaryData.size
        )
        return@withContext bitmap.asImageBitmap()
      }
      return@withContext null
    } catch (e: Exception) {
      Timber.w(e, "extractAlbumArt")
      return@withContext null
    }
  }
}
